File: Modix.Frontend/src/models/PromotionCampaign.ts
import { Dictionary } from "lodash";
import * as _ from "lodash";

import PromotionComment from "@/models/PromotionComment";
import PromotionUserInfo from '@/models/PromotionUserInfo';

export type PromotionSentiment = "For" | "Against" | "Neutral";  
export type CampaignStatus = "Active" | "Approved" | "Denied";

export const SentimentIcons = 
{
    "For": "👍",
    "Against": "👎",
    "Neutral": "😐"
}

export const StatusIcons = 
{
    "Active": "🗳️",
    "Approved": "✔️",
    "Denied": "❌"
}

export default class PromotionCampaign
{
    promotionCampaignId: number = 0;
    promotionFor: PromotionUserInfo = new PromotionUserInfo();

    startDate: Date = new Date(0);
    comments: PromotionComment[] = [];
    status: CampaignStatus = "Active";
    
    get sentimentRatio(): number
    {
        return this.votesFor / (this.votesFor + this.votesAgainst);
    }

    get votesFor(): number
    {
        return this.sentimentCounts['For'] || 0;
    }

    get votesAgainst(): number
    {
        return this.sentimentCounts['Against'] || 0;
    }

    get sentimentCounts(): Dictionary<number>
    {
        return _.countBy(this.comments, (comment: PromotionComment) => comment.sentiment);
    }

    deserializeFrom(input: any): PromotionCampaign
    {
        var self: any = this;
        
        for (let prop in input)
        {
            self[prop] = input[prop];
        }

        return this;
    }
}

File: Modix.Data/Models/PromotionCampaign.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace Modix.Data.Models
{
    public enum CampaignStatus
    {
        Active,
        Approved,
        Denied
    }

    public class PromotionCampaign
    {
        [Key, Required, DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long PromotionCampaignId { get; set; }

        [Required]
        public DiscordUser PromotionFor { get; set; }

        public DateTimeOffset StartDate { get; set; }
        public CampaignStatus Status { get; set; }

        [NotMapped]
        public int ForVotes => Comments.Count(d => d.Sentiment == PromotionSentiment.For);

        [NotMapped]
        public int AgainstVotes => Comments.Count(d => d.Sentiment == PromotionSentiment.Against);

        [NotMapped]
        public int TotalVotes => Comments.Count(d => d.Sentiment != PromotionSentiment.Neutral);

        [NotMapped]
        public float VoteRatio => (float) ForVotes / (ForVotes + AgainstVotes);

        [NotMapped]
        public string SentimentIcon
        {
            get
            {
                if (VoteRatio > 0.67) return "👍";

                return VoteRatio > 0.34 ? "😐" : "👎";
            }
        }

        public DbSet<PromotionComment> Comments { get; set; }
    }
}